SHIP-0023: Add SHIP for BuildRun Status Results (#13)
shipwright-io/community: ships/0006-surface-results-in-buildruns.md

```yaml
spec:
  results:
    - name: a-strategy-result
      description: a-meaningful-description
    - name: another-strategy-result
      description: a-meaningful-description
  buildSteps:
    ...

```

### Build Results

These are results defined at runtime by the BuildRun controller. In this proposal we categorize them in two:

- sources
- output

Following what was mentioned in [EP removal of tekton resources](https://github.com/shipwright-io/build/blob/main/docs/proposals/removal-tekton-resources.md#3-tekton-results), we need to consider the following:

#### Current State of Build Results for Sources

Assuming the following Build definition:

```yaml
apiVersion: shipwright.io/v1alpha1
kind: Build
metadata:
  name: a-build
spec:
  source:
    url: https://github.com/shipwright-io/sample-go

```

We will have the following:

| Result Name                       | Naming Convention                      | Emitted Today | Definition  | Notes                                |
| --------------------------------: | ------------------------------------------: | -------: | ----------: | -----------------------------------: |

> | shp-source-default-commit-sha     |  `shp-source-${sourceType}-${resultName}`   | yes     |  at runtime | `${sourceType}` is `default` as is the only supported git source today |

Reviewer: For consistency, I would call it name here because the type would be `git`.

```suggestion
| shp-source-default-commit-sha     |  `shp-source-${sourceName}-${resultName}`   | yes     |  at runtime | `${sourceName}` is `default` as `spec.source` has no name. |
```

You can refer to https://github.com/shipwright-io/build/blob/main/docs/proposals/removal-tekton-resources.md#3-tekton-results for the naming logic.

| --------------------------------: | ------------------------------------------: |-------: | ----------: |
| shp-source-a-git-source-commit-sha     |  `shp-source-${spec.sources[0].name}-${resultName}` | no     |  at runtime |
| shp-source-a-git-source-commit-author     |  `shp-source-${spec.sources[0].name}-${resultName}` | no     |  at runtime |
| shp-source-an-http-source-size     |  `shp-source-${spec.sources[1].name}-${resultName}` | no     |  at runtime |

#### Build Results Sources API Proposal

Because of the future adoption of sources, the proposed API for `Build Results` should consider the following:

- The _.status_ subresource should provide a relationship between a source and the results emitted in it´s step definition.
- The _.status_ subresource should be able to surface results emitted from different sources.

Therefore the **proposal** is to introduce `_.status.sources_` subresource which holds a list of `SourcesResult`:

```go
type KeyValue map[string]string

type Sources struct {
  Results []SourceResult `json:"sources"`
}

type SourcesResult struct {
  Name   string   `json:"name"`
  Git    KeyValue `json:"git,omitempty"`

>   Http   KeyValue `json:"http,omitempty"` // A new field can be always introduced to support new types, e.g. bundle

Reviewer: I am in favor of strongly typing this for three reasons:

1. We know which results we produce.
2. `kubectl explain buildrun.status.sources.git` will tell you that there is a commit-sha which is only possible if we explicitly define this.
3. We are not forced to use `string` as type. The size of an HTTP source for example would be numeric.

I also think that you need to make it a reference. At least I cannot imagine what `omitempty` would do for an object type (= what makes it empty).

Reviewer: To clarify what I meant:

```go
type GitSourceResult struct {
  CommitSha string `json:"commit-sha"`
}

type HttpSourceResult struct {
  Size int `json:"size"`
}

type SourceResult struct {
  Name   string   `json:"name"`
  Git    *GitSourceResult  `json:"git"`
  Http   *HttpSourceResult `json:"http"` // A new field can be always introduced to support new types, e.g. bundle
}

type BuilRunStatus struct {
  Sources []SourceResult `json:"sources"`
}
```

}
```

A BuildRun will surface these results as follows:

```yaml
status:
  sources:
  - name: default
    git:
      commit-sha: abcd
```

#### Build Results Output API Proposal

Today, we only have a single output step in our strategies. There is no plan to modify this, therefore we propose here to introduce the _.status.output_ subresource in a BuildRun. This will require the following struct:

```go
type Output struct {

>   Results KeyValue `json:"output"`

Reviewer: Same here, I'd love it strongly-typed.

like to surface. All they need to do is:

- Define the result in their strategy, under `spec.results`, this consist of a name and a description.
- Populate or Emit the result to `'$(results.<result-name>.path)'`

#### As a Shipwright Build Developer, I want to surface Build meaningful results

Shipwright Developers should conclude on which results defined at runtime will belong under `.status.sources` and `.status.output`. Today, we emit the following:

- shp-image-size
- shp-image-digest
- shp-source-default-commit-sha

therefore we should populate `.status` subresource with the following:

```yaml
 status:
   sources:
   - name: default
     git:
       commit-sha: cf156cfb0da9be144a40c7d5a110487e5c
   output:
     image-digest: sha:30213102
```

### As a Shipwright CLI developer, I want to consume surfaced results

Workflows on top of Shipwright should be able to consume higher-level data and present that in different ways. As long as they consume from `.status.sources`, `.status.output` or `.status.results`.

### Test Plan

It should include:

- Integration-tests
- Unit-tests

### Release Criteria

Should be available before release v1.0.0

### Risks and Mitigations

> - `Strategy Results` could pile up, ending-up in heavy BuildRun objects.

Reviewer: We will need to document the length limitation behind this which is that Tekton uses container termination messages to transport those values out of a container into a custom resource. And termination messages have a length limitation.

The 4k mentioned here: https://github.com/tektoncd/pipeline/blob/main/docs/tasks.md#emitting-results

- `Build Results` should consider the future adoption of new types, like the `bundle` [one](https://github.com/shipwright-io/build/blob/main/docs/proposals/enable-local-source-code-support.md) or the `local-type` [one](https://github.com/shipwright-io/community/pull/11/).

> - Categorizing `Build Results` per type name might be complex, as we will consume results directly from Tekton, where there is no clear indication of their related step, as they come in the form of key/values. Therefore we need to stick to the `shp-source-${sourceType}-${resultName}` naming convention when defining results at runtime.

Reviewer: ```suggestion
- Categorizing `Build Results` per type name might be complex, as we will consume results directly from Tekton, where there is no clear indication of their related step, as they come in the form of key/values. Therefore we need to stick to the `shp-source-${sourceName}-${resultName}` naming convention when defining results at runtime.
```
